---
title: "R Notebook"
output: html_notebook
---
Messy scripts for plankton data to go from Google Sheets to BCO-DMO submittable table
>NOTE: don't exclude non-snail columns

##libraries
Copy and pasted list of packages used from other Snail BCO-DMO submission; will adjust them as necessary
```{r}
package_list<-c(#"taxize", #NOTE WITH TAXIZE: get_wormsid and similar functions that retrieve from WoRMS using scientific name search marine taxa only UNLESS you install the development version
                "worrms",
                #"plyr", #maybe don't need; if used, need to load before deplyr/tidyverse
                "tidyr",
                "dplyr",
                "tibble")
#install.packages("taxize") #install packages as necessary
#install.packages("tidyr") # "
#install.packages("dplyr") # "
#install.packages("tibble") # "
for (i in package_list) {
  library(i, character.only = TRUE)
}

#development version of taxize
# install.packages("remotes")
# remotes::install_github("ropensci/taxize")
library(taxize)
```
##Cleaning spreadsheet
###FIRST: Download as .csv, load it
https://docs.google.com/spreadsheets/d/1LY00EGjE-voEqPWtJO3-1vVkgXx5aB0TG3wLFGF6a4c/edit#gid=2017687852
   downloaded so we can work with it offline, and not mess with the original copy

Modifications to dataset before loading- see original sheet for context (NOTE: currently modified in Excel b/c convenience; if issues arise from Excel having trouble with micro as a special character, will try modifying it in a text editor)
  removed rows 1-8; data not relevant to the final submission for this (eg, pump name, vol filtered, etc); as row "size fraction" also identifies the mooring number and site name, the mooring and site rows are unnecessary, and the values in row "size fraction" make adaquate headers for the counts columns- they identify the mooring number, site name, and size fraction
  changed the text "size fraction"  to "group" so it may act as a column header for the far left column
  added numbers totals row to reflect mooring number, which in turn reflects site (the mooring number for the counts by site are in the column headers for that site)

Be aware that, when R reads in column headers that start with a number, it automatically adds a leading X, so that, presumably, the column header isn't interpreted as something you can do math on.

```{r}
setwd("C:/Users/Dale/Desktop") #wherever the csv is saved
initial<-read.csv("Mariana_2010_pumps_WORKING-COPY - larval_counts.csv") #csv name

initial_cleaning<-initial #create a duplicate of initial called initial_df so that if I screw something up later, I still have the original df untouched

#save various columns as class character
columns<-colnames(initial_cleaning) #make list of column names to check which ones should be characters
char_cols<-columns[1:9] #columns 1 thru 9 contain character inputs; rest are totals
for (i in 1:length(char_cols)){ #convert columns that should be characters to class "character"
  initial_cleaning[,i]<-as.character(initial_cleaning[,i])
}
```

###SECOND: Clean and check identifications
Add AphiaIDs from Scientific Name
Check taxon rank (from aphiaID vs manual)
Add phylum (manual verification- correct phylum for each entry)
```{r}
##add column to hold AphiaID
initial_cleaning<-add_column(initial_cleaning,AphiaID="",.after="Group")

##add column to hold taxon rank check
initial_cleaning<-add_column(initial_cleaning,taxonRank_AphiaID_match="",.after="taxonRank")

##finding AphiaID from column scientificName
for (i in 1:length(initial_cleaning$scientificName)) {
  initial_cleaning$AphiaID[i]<-wm_name2id(initial_cleaning$scientificName[i])
}
#make AphiaID numeric, for later calls
initial_cleaning$AphiaID<-as.numeric(initial_cleaning$AphiaID)

#check if AphiaID/scientificName match given taxon
##initial run: lots of FALSEes b/c upper/lowercase mismatch; worms all lowercase; changing taxonRank to all lower
initial_cleaning$taxonRank<-tolower(initial_cleaning$taxonRank)
##for each value in taxonRank, see if equal to the rank of the AphiaID; place this true/false as value for taxonRank_AphiaID_match column
for (i in 1:length(initial_cleaning$scientificName)) {
  initial_cleaning$taxonRank_AphiaID_match[i]<-isTRUE(initial_cleaning$taxonRank[i] == as.character(tax_rank(initial_cleaning$AphiaID[i], db="worms")))
}

#get phylum from AphiaID, for manual validation
##add column to hold phylum
initial_cleaning<-add_column(initial_cleaning,Phylum="",.before="scientificName")
##
for (i in 1:length(initial_cleaning$AphiaID)){
  tree<-classification(initial_cleaning$AphiaID[i], db='worms') #get taxonomic stuff from WoRMS based on AphiaID
  df_tree<-as.data.frame(tree[[1]]) #take just the name/rank/id portion, as a df, from classification() output
  if (length(df_tree$name[(which(df_tree$rank=="Phylum"))])==0) { #some are only to Animalia, and so have no value for Phylum in the output from classification(); put "NA" here. This if/else helps prevent the for loop from throwing an error and stopping when it encounters a classification() output that has no phylum
    initial_cleaning$Phylum[i]<-"NA"
    #paste(c("error row:",i," AphiaID:",initial_cleaning$AphiaID[i]), sep="")
  }
  else
    initial_cleaning$Phylum[i]<-df_tree$name[which(df_tree$rank == "Phylum")] #set name of the phylum to initial_cleaning$Phylum value
}

```

###THIRD: Add additional identification data
Generate and insert LSID from AphiaID
```{r}
#using LSID as scientificNameID
#LSID is just a consistent 'preamble' URI, plus AphiaID
for (i in 1:length(initial_cleaning$AphiaID)) {
  initial_cleaning$scientificNameID[i]<-paste("urn:lsid:marinespecies.org:taxname:",initial_cleaning$AphiaID[i],sep="")
}
```


###FOURTH: remove columns that are irrelevant at this point, rectify naming of any we're keeping (as necessary)
Removing:
-Groups column

Renaming:
-Specimen.ID.for.those.with.genetic.sequences to associatedSequences
-Photograph.for.those.without.genetic.seq..bc.photo.submitted.with.genetic.seq. to associatedMedia
-Genetic.notes  to identificationRemarks
-X#.Total's to X#.[SITE].totals's

Notes: a/p/n in "counts" of copepods and ostracods are absent/present/not applicable, respectively
```{r}
#initial_cleaning_2<-initial_cleaning #another backup incase of mistakes
#initial_cleaning<-initial_cleaning_2 #load backup

#check column names, to see names of which we'd like to keep/drop
#colnames(initial_cleaning)

#remove undesired names
initial_cleaning<-initial_cleaning[,-c(1)] #subset initial_cleaning without undesired columns; just removing "Group" column

#rename some (but not all!) of the non-DarwinCore-compliant column names
initial_cleaning <- initial_cleaning %>%
  rename(
    associatedSequences = Specimen.ID.for.those.with.genetic.sequences, #specimenID =/= associated sequence, but presumably a link to the associated sequence can be gotten pretty easily from the specimen ID
    associatedMedia = Photograph.for.those.without.genetic.seq..bc.photo.submitted.with.genetic.seq., #photo name, needs static reference
    identificationRemarks = Genetic.Notes, #probably the closest fit of catagory names for notes on genetic match
    X1.Snail.total = X1.Total, #just to make it clearer which total this refers to
    X2.Snail.total = X2.Total, #also to make it clearer
    X3.Archaean.total = X3.Total,
    X4.Archaean.total = X4.Total,
    X5.Urashima.total = X5.Total,
    X6.Urashima.total = X6.Total
    
  )
```


###FIFTH: Work on Observance table - Presence/Absence columns (occurenceStatus)
NOTE:
The occurrence table SHOULD have duplicate scientificName entries; different rows can and do have different genetic sequencings
The occurrence table SHOULD ALSO have all moorings and site totals and counts, not just Snail

Adding presence/absence info- currently unsure if we want presence/absence at each mooring, site, or just overall. Currently, doing each mooring ->each site ->overall, and can remove unnecessary columns later. For readability, each site is done in a separate code chunk. Each mooring in that site has a separate for() loop. As some row entries have numerical values for counts and a numerical total, while others have p, a, or n (present, absent, or not applicable) in counts columns and an "NA" in the total column, said for() loop assigns "present" to any row with either "p" in one of the counts columns, or a value greater than 0 in the totals column, and "absent" to rows that have neither.

Create the columns for the presence/absence info:
-occurrence at each mooring will be in the columns occurrenceStatus_X#.[SiteName], and at each site in occurrenceStatus_[SiteName]
-occurrenceStatus_[SiteName] summarizes the presence/absence data from the occurrenceStatus_X#.[SiteName] entries for that site, giving the overall presence/absence at that site
-the OVERALL occurrence status of each row entry is summarized in occurrenceStatus_overall
-none of these are darwincore complaint terms, but can be removed/modified to be so later
```{r}
#create occurrence columns, 1 for each Snail site; can condense into 1, but for now keep separate
obs_table<-add_column(initial_cleaning,	occurrenceStatus_X1.Snail="",occurrenceStatus_X2.Snail="",occurrenceStatus_Snail="",occurrenceStatus_X3.Archaean="",occurrenceStatus_X4.Archaean="",occurrenceStatus_Archaean="",occurrenceStatus_X5.Urashima="",occurrenceStatus_X6.Urashima="",occurrenceStatus_Urashima="",occurenceStatus_overall="",.after="identificationRemarks")
```

Snail Vent presence/absence (X1.Snail[...] and X2.Snail[...] columns)
```{r}
#fill in presence/absence based on totals column>0 or counts column presence/absence/NA
##separate for loops for moorings 1 and 2, just for clarity/simplicity
##data is such that either the "total" column has a value and counts columns lack the letters p/a/n, OR the "total" column is NA and counts columns have p, a, or n

#presence/absence for mooring 1 at Snail Vent
for (i in 1:length(obs_table$scientificName)) {
  obsStatus<-paste(obs_table$X1.Snail.300[i],obs_table$X1.Snail.63[i],obs_table$X1.Snail.ship[i],sep="") #create a value to hold all the entries for individual counts (aka step 1 for dealing with columns with a mix of character and numeric values)
    if (isTRUE(grepl("p", obsStatus))==TRUE) { #if the compilation of counts for the site contains the letter p, occurrence= "present"
      obs_table$occurrenceStatus_X1.Snail[i]<-"present" #set occurrence at Snail 1 to "present" if "p" anywhere in that row's
  } else if (isTRUE(as.numeric(obs_table$X1.Snail.total[i])>0) ==TRUE) { #if "p" is not in the counts columns, but the total is >0
      obs_table$occurrenceStatus_X1.Snail[i]<-"present" #set occurrence at Snail 1 to "present"
  } else
      obs_table$occurrenceStatus_X1.Snail[i]<- "absent" #for all others (no 'p' in counts, and total is not >0), mark as absent
}

#presence/absence for mooring 2 at Snail Vent
for (i in 1:length(obs_table$scientificName)) {
  obsStatus<-paste(obs_table$X2.Snail.300[i],obs_table$X2.Snail.63[i],obs_table$X2.Snail.ship[i],sep="")
    if (isTRUE(grepl("p", obsStatus))==TRUE) {
      obs_table$occurrenceStatus_X2.Snail[i]<-"present"
  } else if (isTRUE(as.numeric(obs_table$X2.Snail.total[i])>0) ==TRUE) {
      obs_table$occurrenceStatus_X2.Snail[i]<-"present"
  } else
      obs_table$occurrenceStatus_X2.Snail[i]<- "absent"
}

#overall at Snail vent
for (i in 1:length(obs_table$scientificName)) {
   if (isTRUE(obs_table$occurrenceStatus_X1.Snail[i] == "absent" && obs_table$occurrenceStatus_X2.Snail[i] == "absent") == TRUE) { #if both moorings at snail are "absent" for the row,
    obs_table$occurrenceStatus_Snail[i]<-"absent" #make snail vent "absent"
 } else
    obs_table$occurrenceStatus_Snail[i] <- "present" #otherwise (present/absent or present/present) make overall snail vent "present"
}
```

Archaean presence/absence (X3.Archaean[...] and X4.Archaean[...] columns)
```{r}
#presence/absence for mooring 3 at Archaean
for (i in 1:length(obs_table$scientificName)) {
  obsStatus<-paste(obs_table$X3.Archaean.300[i],obs_table$X3.Archaean.63[i],obs_table$X3.Archaean.ship[i],sep="")
    if (isTRUE(grepl("p", obsStatus))==TRUE) { 
      obs_table$occurrenceStatus_X3.Archaean[i]<-"present"
  } else if (isTRUE(as.numeric(obs_table$X3.Archaean.total[i])>0) ==TRUE) { 
      obs_table$occurrenceStatus_X3.Archaean[i]<-"present"
  } else
      obs_table$occurrenceStatus_X3.Archaean[i]<- "absent"
}

#presence/absence for mooring 4 at Archaean
for (i in 1:length(obs_table$scientificName)) {
  obsStatus<-paste(obs_table$X4.Archaean.300[i],obs_table$X4.Archaean.63[i],obs_table$X4.Archaean.ship[i],sep="")
    if (isTRUE(grepl("p", obsStatus))==TRUE) {
      obs_table$occurrenceStatus_X4.Archaean[i]<-"present"
  } else if (isTRUE(as.numeric(obs_table$X4.Archaean.total[i])>0) ==TRUE) {
      obs_table$occurrenceStatus_X4.Archaean[i]<-"present"
  } else
      obs_table$occurrenceStatus_X4.Archaean[i]<- "absent"
}

#overall at Archaean (if present/present or absent/absent, look for "present")
for (i in 1:length(obs_table$scientificName)) {
    if (isTRUE(obs_table$occurrenceStatus_X3.Archaean[i] == "absent" && obs_table$occurrenceStatus_X4.Archaean[i] == "absent") == TRUE) {
      obs_table$occurrenceStatus_Archaean[i]<-"absent"
  } else
    obs_table$occurrenceStatus_Archaean[i] <- "present" #otherwise (present/absent) is "present" in overall column
}
```

Urashima presence/absence (X5.Urashima[...] and X6.Urashima[...] columns)
```{r}
#presence/absence for mooring 3 at Urashima
for (i in 1:length(obs_table$scientificName)) {
  obsStatus<-paste(obs_table$X5.Urashima.300[i],obs_table$X5.Urashima.63[i],obs_table$X5.Urashima.ship[i],sep="")
    if (isTRUE(grepl("p", obsStatus))==TRUE) { 
      obs_table$occurrenceStatus_X5.Urashima[i]<-"present"
  } else if (isTRUE(as.numeric(obs_table$X5.Urashima.total[i])>0) ==TRUE) { 
      obs_table$occurrenceStatus_X5.Urashima[i]<-"present"
  } else
      obs_table$occurrenceStatus_X5.Urashima[i]<- "absent"
}

#presence/absence for mooring 4 at Urashima
for (i in 1:length(obs_table$scientificName)) {
  obsStatus<-paste(obs_table$X6.Urashima.300[i],obs_table$X6.Urashima.63[i],obs_table$X6.Urashima.ship[i],sep="")
    if (isTRUE(grepl("p", obsStatus))==TRUE) {
      obs_table$occurrenceStatus_X6.Urashima[i]<-"present"
  } else if (isTRUE(as.numeric(obs_table$X6.Urashima.total[i])>0) ==TRUE) {
      obs_table$occurrenceStatus_X6.Urashima[i]<-"present"
  } else
      obs_table$occurrenceStatus_X6.Urashima[i]<- "absent"
}

#overall at Urashima (if present/present or absent/absent, look for "present")
for (i in 1:length(obs_table$scientificName)) {
   if (isTRUE(obs_table$occurrenceStatus_X5.Urashima[i] == "absent" && obs_table$occurrenceStatus_X6.Urashima[i] == "absent") == TRUE) {
      obs_table$occurrenceStatus_Urashima[i]<-"absent"
 } else
  obs_table$occurrenceStatus_Urashima[i] <- "present" #otherwise (present/absent) is "present" in overall column
}
```

Use the above to determine occurrence status overall
```{r}
#if any of the site columns are present, make overall present; otherwise, make absent
##originall, was if all are abesnt, make absent, otherwise, make present; this version accounts for blanks, making overall absent if there's an absent or two and a blank or two
for (i in 1:length(obs_table$scientificName)) {
  if (isTRUE(obs_table$occurrenceStatus_Snail[i] == "present" || obs_table$occurrenceStatus_Archaean[i] == 
             "present" || obs_table$occurrenceStatus_Urashima[i] == "present") ==TRUE ) {
    obs_table$occurenceStatus_overall[i] <- "present"
  } else
    obs_table$occurenceStatus_overall[i] <- "absent"
}
```



##HERE BE DRAGONS
-testing ground to figure out how to get additional steps done
-unfinished, possibly nonfunctional, scripts

###undetermined number - one row per "higher level taxon"
Collapse table such that duplicates are represented in one row
-all genetic squencing, photos, etc, need to be collapsed too, and not lost

testing solution from https://markhneedham.com/blog/2015/06/27/r-dplyr-squashing-multiple-rows-per-group-into-one/ - error: Column `AphiaID` must be length 1 (a summary value), not 2
```{r}
collapsed_cleaning<-initial_cleaning #establish a dataframe to hold collapsed data

#reorganize for aesthetics
collapsed_cleaning<-collapsed_cleaning %>% 
  group_by(AphiaID) %>% #group by AphiaID
    arrange(final.category.used.for.abundance, .by_group=TRUE) #arrange first by AphiaID, then by final.category.used.for.abundance

#summarize within a group_by; using final.category.used.for.abundance, because some disparate entries share aphiaIDs (eg, various gastropods)

collapsed<-collapsed_cleaning %>% #new df called collaped
  group_by(final.category.used.for.abundance) %>% #group by final.category.used.for.abundance
    summarize( #collapse by group
      AphiaID = paste(unique(AphiaID), sep=", "), #aphiaID stays the same; should only be 1 AphiaID per group, so this should work?
      scientificNameID = paste(unique(scientificNameID)), #same
      Phylum = paste(unique(Phylum)),
      taxonRank = paste(unique(taxonRank)),
      taxonRank_AphiaID_match = paste(unique(taxonRank_AphiaID_match)),
      original.category..gastropods.organized.by.size..Bethany.s.benthic.checklist.gastropods.by.genus. = paste(unique(original.category..gastropods.organized.by.size..Bethany.s.benthic.checklist.gastropods.by.genus., collapse=", ")), #paste unique entries of this column into one box; trying not to lose data
      associatedSequences = paste(associatedSequences, collapse=", "), #paste all sequencies for a given group in sequences column of resulting row 
      associatedMedia = paste(associatedMedia, collapse=", "), #paste all media for a given group in media column of resulting row
      identificationRemarks = paste(unique(identificationRemarks, collapse=", ")), #paste all unique ID remarks
      
  #for counts/totals columns, using paste bc of rows with letters in them, hopefully can come back and add/otherwise fix this in another for loop or something
      X1.Snail.300 = paste(X1.Snail.300,collapse="+"),
      X1.Snail.63 = paste(X1.Snail.63,collapse="+"),
      X1.Snail.ship = paste(X1.Snail.ship,collapse="+"),
      X2.Snail.300 = paste(X2.Snail.300,collapse="+"),
      X2.Snail.63 = paste(X2.Snail.63,collapse="+"),
      X2.Snail.ship = paste(X2.Snail.ship,collapse="+"),
      X3.Archaean.300 = paste(X3.Archaean.300,collapse="+"),
      X3.Archaean.63 = paste(X3.Archaean.63,collapse="+"),
      X3.Archaean.ship = paste(X3.Archaean.ship,collapse="+"),
      X4.Archaean.300 = paste(X4.Archaean.300,collapse="+"),
      X4.Archaean.63 = paste(X4.Archaean.63,collapse="+"),
      X4.Archaean.ship = paste(X4.Archaean.ship,collapse="+"),
      X5.Urashima.300 = paste(X5.Urashima.300,collapse="+"),
      X5.Urashima.63 = paste(X5.Urashima63,collapse="+"),
      X5.Urashima.ship = paste(X5.Urashima.ship,collapse="+"),
      X6.Urashima.300 = paste(X6.Urashima.300,collapse="+"),
      X6.Urashima.63 = paste(X6.Urashima63,collapse="+"),
      X6.Urashima.ship = paste(X6.Urashima.ship,collapse="+"),
      X1.Snail.total = paste(X1.Snail.total, collapse="+"),
      X2.Snail.total = paste(X2.Snail.total, collapse="+"),
      X3.Archaean.total = paste(X3.Archaean.total, collapse="+"),
      X4.Archaean.total = paste(X4.Archaean.total, collapse="+"),
      X5.Urashima.total = paste(X5.Urashima.total, collapse="+"),
      X6.Urashima.total = paste(X6.Urashima.total, collapse="+")
    )

```

testing solution from https://www.biostars.org/p/167028/ -ddply in plyr, loading plyr causes issues with rename() earlier
```{r}
collapsed_cleaning<-initial_cleaning[,1:10] #test with just a few cols

collapsed<-ddply(collapsed_cleaning, .(final.category.used.for.abundance), summarize,
                 AphiaID = paste(unique(AphiaID), collapse="+"),
                 scientificNameID = paste(unique(scientificNameID), collapse="+"),
                 Phylum = paste(unique(Phylum), collapse="+"),
                 scientificName = paste(unique(scientificName), collapse="+"),
                 taxonRank = paste(unique(taxonRank), collapse="+"),
                 taxonRank_AphiaID_match = paste(unique(taxonRank_AphiaID_match), collapse="+"),
                 original.category..gastropods.organized.by.size..Bethany.s.benthic.checklist.gastropods.by.genus. = paste(unique(original.category..gastropods.organized.by.size..Bethany.s.benthic.checklist.gastropods.by.genus.)),
                 associatedSequences = paste(unique(associatedSequences), collapse="+"),
                 associatedMedia = paste(unique(associatedMedia), collapse="+")
                )
```

testing solution from https://stackoverflow.com/questions/51516632/aggregate-character-variables, combined with previous solutions re: summarize
-This creates a df with 1 of each AphiaID-final.category.used.for.abundance, but results in multiple entries for, say, Phylum in some places
```{r}
#does sort of what we're looking for, but results in duplicates in ea. column
#collapse3<- intital_cleaning %>%
  #group_by(AphiaID, final.category.used.for.abundance) %>%
 # summarise_all(toString)

#make a df arranged by AphiaID, then final.category, to make it easier to compare to collapsed dataframe and make sure things aren't lost or something
collapsed_forcomparison<-initial_cleaning %>%
  group_by(AphiaID) %>% #group by AphiaID
    arrange(final.category.used.for.abundance, .by_group=TRUE)

#Creating collapsed dataframe - no duplicates in phylum, etc
collapsed<-initial_cleaning%>%
  group_by(AphiaID, final.category.used.for.abundance) %>% #some have same final category and diff aphiaIDs, and vice versa?
  summarise(
    scientificNameID = toString(unique(scientificNameID)), #insert as string, no duplicates
    phylum = toString(unique(Phylum)), #string, no duplicates
    scientificName = toString(unique(scientificName)), 
    original.category..gastropods.organized.by.size..Bethany.s.benthic.checklist.gastropods.by.genus. = toString(unique(original.category..gastropods.organized.by.size..Bethany.s.benthic.checklist.gastropods.by.genus.)),
    associatedSequences = toString(associatedSequences), #ok if duplicates
    associatedMedia = toString(associatedMedia),
    identificationRemarks = toString(identificationRemarks),
  #can't just do sum b/c p/a/n in some rows
    X1.Snail.300 = paste(X1.Snail.300,collapse="+"),
    X1.Snail.63 = paste(X1.Snail.63,collapse="+"),
    X1.Snail.ship = paste(X1.Snail.ship,collapse="+"),
    X2.Snail.300 = paste(X2.Snail.300,collapse="+"),
    X2.Snail.63 = paste(X2.Snail.63,collapse="+"),
    X2.Snail.ship = paste(X2.Snail.ship,collapse="+"),
    X3.Archaean.300 = paste(X3.Archaean.300,collapse="+"),
    X3.Archaean.63 = paste(X3.Archaean.63,collapse="+"),
    X3.Archaean.ship = paste(X3.Archaean.ship,collapse="+"),
    X4.Archaean.300 = paste(X4.Archaean.300,collapse="+"),
    X4.Archaean.63 = paste(X4.Archaean.63,collapse="+"),
    X4.Archaean.ship = paste(X4.Archaean.ship,collapse="+"),
    X5.Urashima.300 = paste(X5.Urashima.300,collapse="+"),
    X5.Urashima.63 = paste(X5.Urashima.63,collapse="+"),
    X5.Urashima.ship = paste(X5.Urashima.ship,collapse="+"),
    X6.Urashima.300 = paste(X6.Urashima.300,collapse="+"),
    X6.Urashima.63 = paste(X6.Urashima.63,collapse="+"),
    X6.Urashima.ship = paste(X6.Urashima.ship,collapse="+"),
    X1.Snail.total = paste(X1.Snail.total, collapse="+"),
    X2.Snail.total = paste(X2.Snail.total, collapse="+"),
    X3.Archaean.total = paste(X3.Archaean.total, collapse="+"),
    X4.Archaean.total = paste(X4.Archaean.total, collapse="+"),
    X5.Urashima.total = paste(X5.Urashima.total, collapse="+"),
    X6.Urashima.total = paste(X6.Urashima.total, collapse="+")
  )

collapsed2<-collapsed #testing for loop

for (i in 1:length(collapsed$AphiaID)) {
  if (isTRUE(grepl("[^a-z]", collapsed[i,]))==FALSE){ #if this row DOESN'T have a letter in it
    collapsed$X1.Snail.300[i]<-eval(collapsed$X1.Snail.300[i]) #do math on this column
  } else
     print ("at least one count contains letters")
}
   
#different for loop for p/a/n;
  #else if (isTRUE(grepl("p"), #COLUMN/ETC - IF THIS ROW-COLUMN HAS AT LEAST 1 P, PUT P; continue this with other else ifs
 
```